src/Orleans/Runtime/GrainReference.cs

```diff
@@ -116,28 +128,39 @@ private GrainReference(GrainId grainId, string genericArgument, SiloAddress syst
         /// </summary>
         /// <param name="other">The reference to copy.</param>
         protected GrainReference(GrainReference other)
-            : this(other.GrainId, other.genericArguments, other.SystemTargetSilo, other.ObserverId) { }
+            : this(other.GrainId, other.genericArguments, other.SystemTargetSilo, other.ObserverId, other.RuntimeClient) { }
 
         #endregion
 
         #region Instance creator factory functions
 
         /// <summary>Constructs a reference to the grain with the specified ID.</summary>
         /// <param name="grainId">The ID of the grain to refer to.</param>
+        /// <param name="runtimeClient">The runtime client</param>
         /// <param name="genericArguments">Type arguments in case of a generic grain.</param>
         /// <param name="systemTargetSilo">Target silo in case of a system target reference.</param>
-        internal static GrainReference FromGrainId(GrainId grainId, string genericArguments = null, SiloAddress systemTargetSilo = null)
+        internal static GrainReference FromGrainId(GrainId grainId, IRuntimeClient runtimeClient, string genericArguments = null, SiloAddress systemTargetSilo = null)
         {
-            return new GrainReference(grainId, genericArguments, systemTargetSilo, null);
+            return new GrainReference(grainId, genericArguments, systemTargetSilo, null, runtimeClient);
         }
 
-        internal static GrainReference NewObserverGrainReference(GrainId grainId, GuidId observerId)
+        internal static GrainReference NewObserverGrainReference(GrainId grainId, GuidId observerId, IRuntimeClient runtimeClient)
         {
-            return new GrainReference(grainId, null, null, observerId);
+            return new GrainReference(grainId, null, null, observerId, runtimeClient);
         }
 
         #endregion
 
+        public void Bind(IGrainFactory factory)
```

Comment:
I should remove this method and add it as an extension method in a separate PR. There's a `#warning` you can search for to see where the runtime client is set during deserialization. Currently, references to foreign clusters will always be deserialized as though they are references to the cluster whose SerializationManager is deserializing it.